// jpf/lib/view/tab-view/index.js
var map = require('../../map');
var CollectionView = require('../collection-view');
var Collection = require('Backbone').Collection;
var TabListView = require('./tab-list-view');
var TabPanelView = require('./tab-panel-view');

function toggleTabPanelViews(views, index) {
  return function (model) {
    var i = views.indexOf(model);

    if (i > 0) {
      toggleTabPanelView(model, i === index + 1);
    }
  }
}

function toggleTabPanelView(model, indexMatch) {
  if (indexMatch) {
    model.get('view').show();
  }
  else {
    model.get('view').hide();
  }
}

function toggleTabListViews(views, index) {
  return function (model) {
    toggleTabListView(model, views.indexOf(model) === index);
  }
}

function toggleTabListView(model, matchesIndex) {
  if (matchesIndex) {
    model.get('view').active();
  }
  else {
    model.get('view').inactive();
  }
}

var TabView = CollectionView.extend({
  tagName: 'div',
  ModelViewClass: TabPanelView,
  switchTab: function (index) {
    this._views.each(toggleTabPanelViews(this._views, index));

    if (this._tabList) {
      this._tabList._views.each(toggleTabListViews(this._tabList._views, index));
    }
  },
  _initializeViews: function (views, comparator) {
    var thisViews = [];
    var self = this;

    this._tabList = new TabListView({
      collection: self._collection,
      app: this._app,
      router: this._router,
      parent: self
    });

    thisViews.push({
      name: null,
      view: this._tabList
    });

    this._collection.each(function (view, index) {
      var panel = new TabPanelView({
        views: new Collection([view]),
        hidden: index !== 0,
        app: this._app,
        router: this._router,
        parent: self
      });
      thisViews.push({name: null, view: panel});
    });

    return new Collection(thisViews);
  },
  _createView: function (name, model) {
    return {
      name: name,
      view: new this.ModelViewClass({
        model: model,
        parent: this,
        app: this._app,
        router: this._router
      })
    };
  }
});

module.exports = TabView;
