// jpf/lib/view/list-view/index.js
var map = require('../../map');
var CollectionView = require('../collection-view');
var View = require('../view');
var Collection = require('Backbone').Collection;
var ToolbarView = require('./toolbar-view');

var ListView = View.extend({
  tagName: 'div',

  CollectionViewClass: CollectionView,
  attributes: null,
  initialize: function (options) {
    options = options || {};
    this._tools = options.tools || [];
    return View.prototype.initialize.call(this, options);
  },
  _initializeViews: function (views, comparator) {
    var thisViews = [];
    var self = this;

    thisViews.push({
      name: null,
      view: new ToolbarView({
        collection: self._tools,
        parent: self,
        app: this._app,
        checkable: false,
        selectable: true
      })
    });

    thisViews.push({
      name: null,
      view: new self.CollectionViewClass({
        collection: self.collection,
        parent: self,
        app: this._app,
        router: this._router,
        checkable: self._checkable
      })
    });

    return new Collection(thisViews);
  }
});

module.exports = ListView;
