// jpf/lib/view/list-view/tab-list-item-view.js
var ModelView = require('../model-view');
var getView = require('../../get-view');
var TabListItemView = ModelView.extend({
  tagName: 'li',
  events: {
    'click': '_onClick',
    'keydown': '_onKeydown'
  },
  attributes: function () {
    return {
      'role': 'tab',
      'id': 'tab-' + this._uuid,
      'aria-controls': 'panel-' + this._uuid,
      'tabindex': this._tabindex
    };
  },
  initialize: function (options) {
    options = options || {};
    this._uuid = options.model.cid;
    this._tabindex = options.tabindex !== undefined ? options.tabindex : -1;
    this._setAttributes(this.attributes());
    this.el.classList.add('tab-' + options.model.get('name').toLowerCase().replace(/[^a-z0-9]+/, '-'));
    return ModelView.prototype.initialize.call(this, options);
  },
  _template: function (model) {
    return model && model.name ? model.name : '';
  },
  render: function () {
    return ModelView.prototype.render.call(this);
  },
  active: function () {
    this._tabindex = 0;
    this.el.setAttribute('tabindex', this._tabindex);
    this.el.focus();
  },
  inactive: function () {
    this._tabindex = -1;
    this.el.setAttribute('tabindex', this._tabindex);
  },
  isActive: function () {
    return this._tabindex === 0;
  },
  _onClick: function () {
    this._parent.switchTab(this._parent._views.map(getView).indexOf(this) % this._parent._views.length);
  },
  _onKeydown: function (event) {
    var right = 39, l = 76, h = 72, left = 37, up = 38, down = 40, j = 74, k = 75, index;

    // Right
    if (event.which === right || event.which === l || event.which === j || event.which === down) {
      index = this._getViewIndex() + 1;
      this._parent.switchTab(index % this._getNumViews());
    }
    // Left
    else if (event.which === left || event.which === h || event.which === k || event.which === up) {
      index = this._getViewIndex() - 1;
      this._parent.switchTab(index < 0 ? this._getNumViews() - 1 : index);
    }
  },
  _getViewIndex: function () {
    return this._parent._views.map(getView).indexOf(this)
  },
  _getNumViews: function () {
    return this._parent._views.length;
  }
});
module.exports = TabListItemView;
